text/0000-ember-data-model-factory-for.md

```diff
@@ -0,0 +1,189 @@
+
+### The contract for `modelFactoryFor`
+
+The return value of `modelFactoryFor` MUST be the result of a call to [`applicationInstance.factoryFor`](https://www.emberjs.com/api/ember/3.4/classes/ApplicationInstance/methods/factoryFor?anchor=factoryFor)
+ where `applicationInstance` is the `owner` returned by using `getOwner(this)` to access the `owner` of the `store` instance.
+
+```typescript
+interface Klass {}
+
+interface Factory {
+  klass: Klass,
+  create(): Klass
+}
+
+interface FactoryMap {
+    [factoryName: string]: Factory
+}
+
+declare function factoryFor<K extends keyof FactoryMap>(factoryName: K): FactoryMap[K];
+
+interface Store {
+  modelFactoryFor(modelName: string): ReturnType<typeof factoryFor>;
+}
+```
+
+Users interested in providing a custom class for their `records` and whom override `modelFactoryFor`,
+ would not need to also change `modelFor`, as this would be the `klass` accessible via the `factory`.
+
+Users wishing to extend the behavior of `modelFactoryFor` could do so in the following manner:
+
+**Example 1:**
+
+**services/store.js**
+```js
+import { getOwner } from '@ember/application';
+import Store from 'ember-data/store';
+
+export default Store.extend({
+  modelFactoryFor(modelName) {
+    if (someCustomCondition) {
+      return getOwner(this).factoryFor(someFactoryName);
+    }
+    
+    return this._super(modelName);
+  }
+});
+```
+ 
+ #### `Model.modelName`
+ 
+ `ember-data` currently sets `modelName` onto the `klass` accessible via the `factory`. For classes that do not
+   inherit from `DS.Model` this would not be done, although end users may do so themselves in their implementations
+   if so desired.
+
+### What is a valid factory?
+
+The default export of a custom ModelClass **MUST** conform to the requirements of `Ember.factoryFor`. The requirements
+ of `factoryFor` are currently underspecified; however, in practice, this means that the default export is an
+ instantiable class with a static `create` method and an instance `destroy` method or that inherits from `EmberObject`
+ (which provides such methods).
+ 
+**Example 2:**
+
+```javascript
+import { assign } from '@ember/polyfills';
+
+export default class CustomModel {
+  constructor(createArgs) {
+    assign(this, createArgs);
+  }
+  destroy() {
+    // ... do teardown
+  }
+  static create(createArgs) {
+    return new this(createArgs);
+  }
+}
+```
+
+**Example 3:**
+
+```javascript
+import EmberObject from '@ember/object';
+
+export default class CustomModel extends EmberObject {
+  constructor(createArgs) {
+    super(createArgs);
+  }
+}
+```
+
+Custom classes for models should expect their constructor to receive a single argument: an object with *at least*
+ the following.
+ 
+- A `recordData` instance available on the `recordData` `Symbol` and accessible via `getRecordData` (see below)
+- Any properties passed as the second arg to `createRecord`
+- An `owner` available on the owner `Symbol` and accessible via `Ember.getOwner`
```

Comment:
I don't think the `Ember.getOwner` docs mention anything about a `Symbol`; let's just say the owner is accessible via `Ember.getOwner`